### PR TITLE
Add varying batch size support for LUCudaSparseSolver

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py
Comment (mhmukadam) [inline theseus/optimizer/nonlinear/nonlinear_optimizer.py:304]: Is the idea to allow changing the batch size after the linear system was first setup? This seems like a nice fix!

I remember @maurimo had mentioned that the LUCudaSparse implementation currently needed to know a fixed batch size at setup time. But if you use the largest size you think you need then you can use any size smaller than that, for example see the [unit test here](https://github.com/facebookresearch/theseus/blob/eb84ce66ec0e550021ddbd0cb21bbd98ca4b09ba/theseus/optimizer/linear/tests/test_lu_cuda_sparse_solver.py#L93).

**Review by mhmukadam — APPROVED** [review]

**Comment by mhmukadam** [comment]
Looks good to me. Should we merge this to main if this is a general fix and then merge into experimental?

**Comment by fantaosha** [comment]
> Looks good to me. Should we merge this to main if this is a general fix and then merge into experimental?

This is a quick fix. Maybe we had better not merge it main at the moment.

File: theseus/optimizer/nonlinear/nonlinear_optimizer.py
Comment (maurimo) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py:304]: This is certainly feasible, but is it a good idea? The reset is an expensive operation, because it's redoing the symbolic decomposition (actually, many symbolic decompositions, as many as contexts we have available - we need several contexts to allow unroll, this is a limitation of Cublas we cannot go around), so I would not make it happen automatically. It's an operation to be done explicity if the user really wants it, not because we encountered an irregular batch. In the previous discussions, the consensus was that probably the user prefers discarding a sporadic irregular batch instead of triggering the recomputation of all symbolic decompositions, then again when we go back to the ordinary batch size.

I think the way around this could be a function of the solver informing the user if the batch size is mandatory (eg `is_batch_size_fixed()`), so the user knows that the irregular batches are better discarded, or if not reset has to be called again. Is there any case where this would not be the best solution?

Comment (fantaosha) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py:304]: I agree this is not a good idea. In fact, this is a quick fix to the experimental branch. @maurimo Can we use smaller batch size after declaring a large one?

Comment (luisenp) [reply theseus/optimizer/nonlinear/nonlinear_optimizer.py:304]: I vaguely recalled some discussion about this point, but could not remember what the conclusion was, so I couldn't advise @fantaosha on this. Thanks for clarifying!

I think it's fine to merge this to `experimental`, but definitely not to `main`. We need to think a bit what would be a user friendly way to flag this. One option could be to add a flag somewhere (unsure what's the best place, TheseusLayer maybe) that specifies if dynamic batch size is allowed, and make it default to `False`. Then we throw an error whenever this happens, and mention that this option can be turned on, possibly at a huge cost in running time. 

PS: We should have a better registry of previous discussions. As the library gets bigger it's becoming harder to keep track of all resolutions :)

**Review by luisenp — APPROVED** [review]
Approving for merging to current experimental.

**Pull request merged**
